File: ghmm/tui.py
"""TUI interface for ghmm using Textual."""

import pyperclip
from textual.app import App, ComposeResult
from textual.containers import Container, Horizontal, Vertical
from textual.widgets import Button, DataTable, Footer, Header, Input, Label, Static
from textual.screen import ModalScreen
from rich.text import Text

from .config import Config
from .ssh import SSHManager
from .shell import ShellManager
from .gitconfig import GitConfigManager


class AddAccountScreen(ModalScreen):
    """Modal screen for adding a new account."""

    def compose(self) -> ComposeResult:
        yield Container(
            Label("Add New GitHub Account", id="dialog-title"),
            Input(placeholder="Account name (e.g., work, personal)", id="input-name"),
            Input(placeholder="GitHub username", id="input-username"),
            Input(placeholder="GitHub email", id="input-email"),
            Input(placeholder="Directory path (e.g., ~/code/work)", id="input-directory"),
            Horizontal(
                Button("Add", variant="success", id="btn-add"),
                Button("Cancel", variant="error", id="btn-cancel"),
                id="dialog-buttons",
            ),
            id="add-account-dialog",
        )

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == "btn-cancel":
            self.app.pop_screen()
        elif event.button.id == "btn-add":
            # Get values from inputs
            name = self.query_one("#input-name", Input).value
            username = self.query_one("#input-username", Input).value
            email = self.query_one("#input-email", Input).value
            directory = self.query_one("#input-directory", Input).value

            if not all([name, username, email, directory]):
                self.app.show_message("Error", "All fields are required!")
                return

            # Add the account
            result = self.app.add_account(name, username, email, directory)
            if result:
                self.app.pop_screen()
            else:
                self.app.show_message("Error", f"Account '{name}' already exists!")


class MessageScreen(ModalScreen):
    """Modal screen for displaying messages."""

    def __init__(self, title: str, message: str):
        super().__init__()
        self.title_text = title
        self.message_text = message

    def compose(self) -> ComposeResult:
        yield Container(
            Label(self.title_text, id="message-title"),
            Static(self.message_text, id="message-body"),
            Button("OK", variant="primary", id="btn-ok"),
            id="message-dialog",
        )

    def on_button_pressed(self, event: Button.Pressed) -> None:
        self.app.pop_screen()


class GHMMApp(App):
    """GitHub Multi-Account Manager TUI application."""

    CSS = """
    Screen {
        background: $surface;
    }

    #add-account-dialog, #message-dialog {
        width: 60;
        height: auto;
        border: thick $primary;
        background: $surface;
        padding: 1;
    }

    #dialog-title, #message-title {
        width: 100%;
        content-align: center middle;
        text-style: bold;
        color: $accent;
        margin-bottom: 1;
    }

    #message-body {
        width: 100%;
        margin-bottom: 1;
    }

    Input {
        margin: 1 0;
    }

    #dialog-buttons {
        width: 100%;
        height: auto;
        align: center middle;
    }

    #main-container {
        padding: 1;
    }

    #accounts-table {
        height: 1fr;
        margin: 1 0;
    }

    #actions {
        height: auto;
        margin: 1 0;
    }

    Button {
        margin: 0 1;
    }

    #status-bar {
        dock: bottom;
        height: 3;
        background: $panel;
        padding: 1;
    }
    """

    BINDINGS = [
        ("q", "quit", "Quit"),
        ("a", "add_account", "Add Account"),
        ("d", "delete_account", "Delete Account"),
        ("t", "test_connection", "Test Connection"),
        ("r", "refresh", "Refresh"),
    ]

    def __init__(self):
        super().__init__()
        self.config = Config()
        self.ssh_manager = SSHManager()
        self.shell_manager = ShellManager()
        self.git_manager = GitConfigManager()

    def compose(self) -> ComposeResult:
        yield Header()
        yield Container(
            Label("📦 GitHub Multi-Account Manager", id="title"),
            DataTable(id="accounts-table"),
            Horizontal(
                Button("➕ Add Account", variant="success", id="btn-add"),
                Button("🗑️ Delete Account", variant="error", id="btn-delete"),
                Button("🔑 Copy SSH Key", variant="primary", id="btn-copy-key"),
                Button("✅ Test Connection", variant="primary", id="btn-test"),
                Button("⚙️ Apply Config", variant="warning", id="btn-apply"),
                id="actions",
            ),
            Static("", id="status-bar"),
            id="main-container",
        )
        yield Footer()

    def on_mount(self) -> None:
        """Initialize the table when the app starts."""
        table = self.query_one(DataTable)
        table.add_columns("Name", "Username", "Email", "Directory", "Status")
        self.refresh_table()
        self.update_status(f"Detected shell: {self.shell_manager.shell_type}")

    def refresh_table(self) -> None:
        """Refresh the accounts table."""
        table = self.query_one(DataTable)
        table.clear()

        accounts = self.config.list_accounts()
        default = self.config.get_default_account()

        for account in accounts:
            # Check if SSH key exists
            from pathlib import Path
            key_exists = Path(account['ssh_key_path']).exists()
            status = "✅ Ready" if key_exists else "⚠️ No Key"

            # Mark default account
            name = account['name']
            if name == default:
                name = f"⭐ {name}"

            table.add_row(
                name,
                account['username'],
                account['email'],
                account['directory'],
                status,
            )

    def action_add_account(self) -> None:
        """Show the add account dialog."""
        self.push_screen(AddAccountScreen())

    def action_delete_account(self) -> None:
        """Delete the selected account."""
        table = self.query_one(DataTable)
        if table.cursor_row < 0:
            self.show_message("Error", "Please select an account to delete")
            return

        accounts = self.config.list_accounts()
        if table.cursor_row >= len(accounts):
            return

        account = accounts[table.cursor_row]
        self.config.remove_account(account['name'])
        self.git_manager.remove_account_gitconfig(account['name'])
        self.refresh_table()
        self.update_status(f"Deleted account: {account['name']}")

    def action_test_connection(self) -> None:
        """Test SSH connection for selected account."""
        table = self.query_one(DataTable)
        if table.cursor_row < 0:
            self.show_message("Error", "Please select an account to test")
            return

        accounts = self.config.list_accounts()
        if table.cursor_row >= len(accounts):
            return

        account = accounts[table.cursor_row]
        success, message = self.ssh_manager.test_connection(account['host_alias'])

        if success:
            self.show_message("✅ Connection Successful", message)
        else:
            self.show_message("❌ Connection Failed", message)

    def action_refresh(self) -> None:
        """Refresh the table."""
        self.refresh_table()
        self.update_status("Refreshed")

    def on_button_pressed(self, event: Button.Pressed) -> None:
        """Handle button presses."""
        if event.button.id == "btn-add":
            self.action_add_account()
        elif event.button.id == "btn-delete":
            self.action_delete_account()
        elif event.button.id == "btn-test":
            self.action_test_connection()
        elif event.button.id == "btn-copy-key":
            self.copy_ssh_key()
        elif event.button.id == "btn-apply":
            self.apply_all_config()

    def copy_ssh_key(self) -> None:
        """Copy SSH public key to clipboard."""
        table = self.query_one(DataTable)
        if table.cursor_row < 0:
            self.show_message("Error", "Please select an account")
            return

        accounts = self.config.list_accounts()
        if table.cursor_row >= len(accounts):
            return

        account = accounts[table.cursor_row]
        pub_key = self.ssh_manager.get_public_key(account['ssh_key_path'])

        if pub_key:
            try:
                pyperclip.copy(pub_key)
                self.show_message("✅ Copied!", f"SSH key copied to clipboard for {account['name']}")
            except Exception as e:
                self.show_message("Error", f"Failed to copy to clipboard: {str(e)}")
        else:
            self.show_message("Error", "SSH key not found. Generate one first!")

    def apply_all_config(self) -> None:
        """Apply all configurations (SSH, git, shell)."""
        accounts = self.config.list_accounts()
        default = self.config.get_default_account()

        # Update SSH config
        success, msg = self.ssh_manager.update_ssh_config(accounts)
        if not success:
            self.show_message("Error", msg)
            return

        # Update gitconfig
        success, msg = self.git_manager.update_gitconfig(accounts)
        if not success:
            self.show_message("Error", msg)
            return

        # Create individual gitconfigs
        for account in accounts:
            self.git_manager.create_account_gitconfig(account)

        # Update shell config
        success, msg = self.shell_manager.update_shell_config(accounts, default)
        if not success:
            self.show_message("Error", msg)
            return

        reload_cmd = self.shell_manager.get_reload_command()
        self.show_message(
            "✅ Configuration Applied!",
            f"All configs updated successfully!\\n\\nReload your shell with:\\n{reload_cmd}"
        )
        self.update_status("Configuration applied successfully")

    def add_account(self, name: str, username: str, email: str, directory: str) -> bool:
        """Add a new account with full setup."""
        # Add to config
        success = self.config.add_account(name, username, email, directory)
        if not success:
            return False

        account = self.config.get_account(name)
        
        # Generate SSH key
        key_success, key_msg = self.ssh_manager.generate_key(
            account['ssh_key_path'],
            email
        )

        if key_success:
            # Add to ssh-agent
            self.ssh_manager.add_to_ssh_agent(account['ssh_key_path'])
            self.update_status(f"Added account: {name} (SSH key generated)")
        else:
            self.update_status(f"Added account: {name} (SSH key generation failed)")

        self.refresh_table()
        return True

    def show_message(self, title: str, message: str) -> None:
        """Show a message dialog."""
        self.push_screen(MessageScreen(title, message))

    def update_status(self, message: str) -> None:
        """Update the status bar."""
        status = self.query_one("#status-bar", Static)
        status.update(message)


def run_tui():
    """Run the TUI application."""
    app = GHMMApp()
    app.run()

File: internal/tui/tui.go
package tui

import (
	"fmt"
	"os"
	"strings"

	"github.com/atotto/clipboard"
	"github.com/charmbracelet/bubbles/key"
	"github.com/charmbracelet/bubbles/table"
	"github.com/charmbracelet/bubbles/textinput"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/donbowman/github-multi-account-manager/internal/config"
	"github.com/donbowman/github-multi-account-manager/internal/git"
	"github.com/donbowman/github-multi-account-manager/internal/shell"
	"github.com/donbowman/github-multi-account-manager/internal/ssh"
)

var (
	titleStyle = lipgloss.NewStyle().
			Bold(true).
			Foreground(lipgloss.Color("205")).
			MarginBottom(1)

	helpStyle = lipgloss.NewStyle().
			Foreground(lipgloss.Color("241")).
			MarginTop(1)

	statusStyle = lipgloss.NewStyle().
			Foreground(lipgloss.Color("green")).
			Bold(true)

	errorStyle = lipgloss.NewStyle().
			Foreground(lipgloss.Color("red")).
			Bold(true)

	infoStyle = lipgloss.NewStyle().
			Foreground(lipgloss.Color("cyan"))

	baseStyle = lipgloss.NewStyle().
			BorderStyle(lipgloss.NormalBorder()).
			BorderForeground(lipgloss.Color("240"))
)

type keyMap struct {
	Up          key.Binding
	Down        key.Binding
	Quit        key.Binding
	Refresh     key.Binding
	Apply       key.Binding
	CopyKey     key.Binding
	ShowDetails key.Binding
	AddAccount  key.Binding
	AutoSync    key.Binding
	GenerateKey key.Binding
	TestConn    key.Binding
}

var keys = keyMap{
	Up: key.NewBinding(
		key.WithKeys("up", "k"),
		key.WithHelp("↑/k", "up"),
	),
	Down: key.NewBinding(
		key.WithKeys("down", "j"),
		key.WithHelp("↓/j", "down"),
	),
	Quit: key.NewBinding(
		key.WithKeys("q", "ctrl+c"),
		key.WithHelp("q", "quit"),
	),
	Refresh: key.NewBinding(
		key.WithKeys("r"),
		key.WithHelp("r", "refresh"),
	),
	Apply: key.NewBinding(
		key.WithKeys("a"),
		key.WithHelp("a", "apply configs"),
	),
	CopyKey: key.NewBinding(
		key.WithKeys("c"),
		key.WithHelp("c", "copy SSH key"),
	),
	ShowDetails: key.NewBinding(
		key.WithKeys("enter", " "),
		key.WithHelp("enter", "details"),
	),
	AddAccount: key.NewBinding(
		key.WithKeys("n"),
		key.WithHelp("n", "add account"),
	),
	AutoSync: key.NewBinding(
		key.WithKeys("s"),
		key.WithHelp("s", "auto-sync"),
	),
	GenerateKey: key.NewBinding(
		key.WithKeys("g"),
		key.WithHelp("g", "generate SSH key"),
	),
	TestConn: key.NewBinding(
		key.WithKeys("t"),
		key.WithHelp("t", "test connection"),
	),
}

type viewMode int

const (
	viewTable viewMode = iota
	viewAddAccount
	viewDetails
)

type model struct {
	table        table.Model
	config       *config.Config
	sshManager   *ssh.Manager
	gitManager   *git.Manager
	shellManager *shell.Manager
	statusMsg    string
	errorMsg     string
	mode         viewMode
	detailsText  string
	// Form fields for adding account
	formInputs   []textinput.Model
	formFocused  int
	emptyStartup bool // True if started with no accounts
}

func (m model) Init() tea.Cmd {
	return nil
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	var cmd tea.Cmd

	switch msg := msg.(type) {
	case tea.KeyMsg:
		// If showing details, any key dismisses
		if m.mode == viewDetails {
			m.mode = viewTable
			m.detailsText = ""
			return m, nil
		}

		// If in add account mode, handle form navigation
		if m.mode == viewAddAccount {
			return m.handleFormInput(msg)
		}

		switch {
		case key.Matches(msg, keys.Quit):
			return m, tea.Quit

		case key.Matches(msg, keys.Refresh):
			m = m.refreshTable()
			m.statusMsg = "✓ Refreshed"
			m.errorMsg = ""

		case key.Matches(msg, keys.Apply):
			m = m.applyConfigs()

		case key.Matches(msg, keys.CopyKey):
			m = m.copySSHKey()

		case key.Matches(msg, keys.ShowDetails):
			m = m.showDetails()

		case key.Matches(msg, keys.AddAccount):
			m = m.startAddAccount()

		case key.Matches(msg, keys.AutoSync):
			m = m.autoSync()

		case key.Matches(msg, keys.GenerateKey):
			m = m.generateSSHKey()

		case key.Matches(msg, keys.TestConn):
			m = m.testConnection()
		}
	}

	m.table, cmd = m.table.Update(msg)
	return m, cmd
}

func (m model) View() string {
	switch m.mode {
	case viewDetails:
		return m.renderDetails()
	case viewAddAccount:
		return m.renderAddAccountForm()
	default:
		return m.renderTable()
	}
}

func (m model) renderTable() string {
	title := titleStyle.Render("📦 GitHub Multi-Account Manager")

	var status string
	if m.errorMsg != "" {
		status = errorStyle.Render(m.errorMsg)
	} else if m.statusMsg != "" {
		status = statusStyle.Render(m.statusMsg)
	}

	help := helpStyle.Render(
		"q:quit • n:add • s:sync • g:gen key • t:test • a:apply • c:copy • enter:details",
	)

	return fmt.Sprintf("%s\n\n%s\n\n%s\n%s\n",
		title,
		baseStyle.Render(m.table.View()),
		status,
		help,
	)
}

func (m model) renderDetails() string {
	title := titleStyle.Render("Account Details")
	content := baseStyle.Render(m.detailsText)
	help := helpStyle.Render("Press any key to return")

	return fmt.Sprintf("%s\n\n%s\n\n%s\n", title, content, help)
}

func (m model) refreshTable() model {
	accounts := m.config.ListAccounts()
	defaultAcc := m.config.GetDefaultAccount()

	rows := []table.Row{}
	for _, acc := range accounts {
		name := acc.Name
		if name == defaultAcc {
			name = "⭐ " + name
		}

		// Check if SSH key exists
		status := "⚠️ No key"
		if _, err := os.Stat(acc.SSHKeyPath); err == nil {
			status = "✅ Ready"
		}

		rows = append(rows, table.Row{
			name,
			acc.Username,
			acc.Email,
			acc.Directory,
			status,
		})
	}

	m.table.SetRows(rows)
	return m
}

func (m model) applyConfigs() model {
	accounts := m.config.ListAccounts()
	defaultAcc := m.config.GetDefaultAccount()

	// Update SSH config
	if err := m.sshManager.UpdateSSHConfig(accounts); err != nil {
		m.errorMsg = fmt.Sprintf("❌ SSH config failed: %v", err)
		m.statusMsg = ""
		return m
	}

	// Update Git config
	if err := m.gitManager.UpdateGitconfig(accounts); err != nil {
		m.errorMsg = fmt.Sprintf("❌ Git config failed: %v", err)
		m.statusMsg = ""
		return m
	}

	// Create individual gitconfigs
	for _, acc := range accounts {
		if err := m.gitManager.CreateAccountGitconfig(acc); err != nil {
			m.errorMsg = fmt.Sprintf("❌ Failed to create gitconfig for %s: %v", acc.Name, err)
			m.statusMsg = ""
			return m
		}
	}

	// Update shell config
	if err := m.shellManager.UpdateShellConfig(accounts, defaultAcc); err != nil {
		m.errorMsg = fmt.Sprintf("❌ Shell config failed: %v", err)
		m.statusMsg = ""
		return m
	}

	m.statusMsg = fmt.Sprintf("✓ Configs applied! Reload shell: %s", m.shellManager.GetReloadCommand())
	m.errorMsg = ""
	return m
}

func (m model) copySSHKey() model {
	if m.table.Cursor() < 0 {
		m.errorMsg = "❌ No account selected"
		m.statusMsg = ""
		return m
	}

	accounts := m.config.ListAccounts()
	if m.table.Cursor() >= len(accounts) {
		return m
	}

	account := accounts[m.table.Cursor()]
	pubKey, err := m.sshManager.GetPublicKey(account.SSHKeyPath)
	if err != nil {
		m.errorMsg = fmt.Sprintf("❌ No SSH key found for %s", account.Name)
		m.statusMsg = ""
		return m
	}

	if err := clipboard.WriteAll(pubKey); err != nil {
		m.errorMsg = "❌ Failed to copy to clipboard"
		m.statusMsg = ""
		return m
	}

	m.statusMsg = fmt.Sprintf("✓ SSH key for %s copied to clipboard!", account.Name)
	m.errorMsg = ""
	return m
}

func (m model) showDetails() model {
	if m.table.Cursor() < 0 {
		return m
	}

	accounts := m.config.ListAccounts()
	if m.table.Cursor() >= len(accounts) {
		return m
	}

	account := accounts[m.table.Cursor()]

	var details strings.Builder
	details.WriteString(fmt.Sprintf("Account: %s\n\n", infoStyle.Render(account.Name)))
	details.WriteString(fmt.Sprintf("Username:     %s\n", account.Username))
	details.WriteString(fmt.Sprintf("Email:        %s\n", account.Email))
	details.WriteString(fmt.Sprintf("Directory:    %s\n", account.Directory))
	details.WriteString(fmt.Sprintf("SSH Key:      %s\n", account.SSHKeyPath))
	details.WriteString(fmt.Sprintf("Host Alias:   %s\n\n", account.HostAlias))

	// Check SSH key status
	if _, err := os.Stat(account.SSHKeyPath); err == nil {
		pubKey, err := m.sshManager.GetPublicKey(account.SSHKeyPath)
		if err == nil {
			details.WriteString(statusStyle.Render("✓ SSH Key exists\n\n"))
			details.WriteString("Public Key:\n")
			details.WriteString(infoStyle.Render(pubKey))
			details.WriteString("\n\nPress 'c' in main view to copy to clipboard")
		}
	} else {
		details.WriteString(errorStyle.Render("⚠ No SSH key found\n"))
		details.WriteString("Generate with: ssh-keygen -t ed25519 -C \"" + account.Email + "\" -f " + account.SSHKeyPath)
	}

	m.detailsText = details.String()
	m.mode = viewDetails
	return m
}

func (m model) startAddAccount() model {
	// Initialize form inputs
	inputs := make([]textinput.Model, 4)

	inputs[0] = textinput.New()
	inputs[0].Placeholder = "work, personal, etc."
	inputs[0].Focus()
	inputs[0].CharLimit = 50
	inputs[0].Width = 40
	inputs[0].Prompt = "Account name: "

	inputs[1] = textinput.New()
	inputs[1].Placeholder = "your-github-username"
	inputs[1].CharLimit = 100
	inputs[1].Width = 40
	inputs[1].Prompt = "GitHub username: "

	inputs[2] = textinput.New()
	inputs[2].Placeholder = "you@example.com"
	inputs[2].CharLimit = 100
	inputs[2].Width = 40
	inputs[2].Prompt = "Email: "

	inputs[3] = textinput.New()
	inputs[3].Placeholder = "~/code/work"
	inputs[3].CharLimit = 200
	inputs[3].Width = 40
	inputs[3].Prompt = "Directory: "

	m.formInputs = inputs
	m.formFocused = 0
	m.mode = viewAddAccount
	m.statusMsg = ""
	m.errorMsg = ""

	return m
}

func (m model) handleFormInput(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "esc":
		m.mode = viewTable
		m.formInputs = nil
		return m, nil

	case "tab", "down":
		m.formInputs[m.formFocused].Blur()
		m.formFocused++
		if m.formFocused >= len(m.formInputs) {
			m.formFocused = 0
		}
		m.formInputs[m.formFocused].Focus()
		return m, nil

	case "shift+tab", "up":
		m.formInputs[m.formFocused].Blur()
		m.formFocused--
		if m.formFocused < 0 {
			m.formFocused = len(m.formInputs) - 1
		}
		m.formInputs[m.formFocused].Focus()
		return m, nil

	case "enter":
		// Validate and save
		name := strings.TrimSpace(m.formInputs[0].Value())
		username := strings.TrimSpace(m.formInputs[1].Value())
		email := strings.TrimSpace(m.formInputs[2].Value())
		directory := strings.TrimSpace(m.formInputs[3].Value())

		if name == "" || username == "" || email == "" || directory == "" {
			m.errorMsg = "❌ All fields are required"
			return m, nil
		}

		// Add account (this also saves the config)
		if err := m.config.AddAccount(name, username, email, directory); err != nil {
			m.errorMsg = fmt.Sprintf("❌ Failed to add account: %v", err)
			return m, nil
		}

		m.mode = viewTable
		m.formInputs = nil
		m = m.refreshTable()
		m.statusMsg = fmt.Sprintf("✓ Added account '%s'! Press 'a' to apply configs", name)
		m.errorMsg = ""

		// If this was the first account during startup, set as default
		if m.emptyStartup && len(m.config.ListAccounts()) == 1 {
			m.config.SetDefaultAccount(name)
		}

		return m, nil
	}

	// Update the focused input
	var cmd tea.Cmd
	m.formInputs[m.formFocused], cmd = m.formInputs[m.formFocused].Update(msg)
	return m, cmd
}

func (m model) renderAddAccountForm() string {
	title := titleStyle.Render("Add New Account")

	var form strings.Builder
	form.WriteString("\n")
	for i, input := range m.formInputs {
		form.WriteString(input.View())
		form.WriteString("\n")
		if i < len(m.formInputs)-1 {
			form.WriteString("\n")
		}
	}

	var status string
	if m.errorMsg != "" {
		status = errorStyle.Render(m.errorMsg)
	}

	help := helpStyle.Render("tab/↓:next • shift+tab/↑:prev • enter:save • esc:cancel")

	return fmt.Sprintf("%s\n\n%s\n\n%s\n%s\n",
		title,
		baseStyle.Render(form.String()),
		status,
		help,
	)
}

func (m model) autoSync() model {
	imported, err := m.config.ImportFromExistingSetup()
	if err != nil {
		m.errorMsg = fmt.Sprintf("❌ Auto-sync failed: %v", err)
		m.statusMsg = ""
		return m
	}

	if imported == 0 {
		m.statusMsg = "No accounts found in .gitconfig or .ssh/config"
		m.errorMsg = ""
	} else {
		m = m.refreshTable()
		m.statusMsg = fmt.Sprintf("✓ Imported %d account(s) from existing setup!", imported)
		m.errorMsg = ""
	}

	return m
}

func (m model) generateSSHKey() model {
	if m.table.Cursor() < 0 {
		m.errorMsg = "❌ No account selected"
		m.statusMsg = ""
		return m
	}

	accounts := m.config.ListAccounts()
	if m.table.Cursor() >= len(accounts) {
		return m
	}

	account := accounts[m.table.Cursor()]

	// Check if key already exists
	if _, err := os.Stat(account.SSHKeyPath); err == nil {
		m.errorMsg = fmt.Sprintf("⚠️  SSH key already exists for %s", account.Name)
		m.statusMsg = ""
		return m
	}

	// Generate the key
	if err := m.sshManager.GenerateKey(account.SSHKeyPath, account.Email, "ed25519"); err != nil {
		m.errorMsg = fmt.Sprintf("❌ Failed to generate key: %v", err)
		m.statusMsg = ""
		return m
	}

	// Get and copy the public key
	pubKey, err := m.sshManager.GetPublicKey(account.SSHKeyPath)
	if err != nil {
		m.errorMsg = fmt.Sprintf("❌ Failed to read public key: %v", err)
		m.statusMsg = ""
		return m
	}

	if err := clipboard.WriteAll(pubKey); err != nil {
		m.statusMsg = fmt.Sprintf("✓ Key generated for %s! Press 'enter' to see details", account.Name)
	} else {
		m.statusMsg = fmt.Sprintf("✓ SSH key generated and copied! Add it to GitHub, then press 't' to test", account.Name)
	}
	m.errorMsg = ""
	m = m.refreshTable()

	return m
}

func (m model) testConnection() model {
	if m.table.Cursor() < 0 {
		m.errorMsg = "❌ No account selected"
		m.statusMsg = ""
		return m
	}

	accounts := m.config.ListAccounts()
	if m.table.Cursor() >= len(accounts) {
		return m
	}

	account := accounts[m.table.Cursor()]

	// Check if SSH key exists
	if _, err := os.Stat(account.SSHKeyPath); err != nil {
		m.errorMsg = fmt.Sprintf("❌ No SSH key found for %s. Press 'g' to generate one", account.Name)
		m.statusMsg = ""
		return m
	}

	m.statusMsg = fmt.Sprintf("Testing connection for %s...", account.Name)

	// Test the connection
	success, message := m.sshManager.TestConnection(account.HostAlias)

	if success {
		m.statusMsg = fmt.Sprintf("✓ %s connected successfully!", account.Name)
		m.errorMsg = ""
	} else {
		m.errorMsg = fmt.Sprintf("❌ Connection failed for %s: %s", account.Name, message)
		m.statusMsg = ""
	}

	m = m.refreshTable()
	return m
}

// Run starts the TUI application
func Run() error {
	// Initialize managers
	cfg, err := config.New("")
	if err != nil {
		return fmt.Errorf("failed to initialize config: %w", err)
	}

	emptyStartup := false

	// Auto-import from existing setup if no accounts configured
	if len(cfg.ListAccounts()) == 0 {
		emptyStartup = true
		if imported, err := cfg.ImportFromExistingSetup(); err == nil && imported > 0 {
			fmt.Printf("✨ Auto-detected %d account(s) from .gitconfig and .ssh/config\n", imported)
			emptyStartup = false
		}
	}

	sshMgr, err := ssh.New()
	if err != nil {
		return fmt.Errorf("failed to initialize SSH manager: %w", err)
	}

	gitMgr, err := git.New()
	if err != nil {
		return fmt.Errorf("failed to initialize Git manager: %w", err)
	}

	shellMgr, err := shell.New()
	if err != nil {
		return fmt.Errorf("failed to initialize shell manager: %w", err)
	}

	// Create table
	columns := []table.Column{
		{Title: "Name", Width: 20},
		{Title: "Username", Width: 20},
		{Title: "Email", Width: 35},
		{Title: "Directory", Width: 30},
		{Title: "Status", Width: 12},
	}

	t := table.New(
		table.WithColumns(columns),
		table.WithFocused(true),
		table.WithHeight(10),
	)

	s := table.DefaultStyles()
	s.Header = s.Header.
		BorderStyle(lipgloss.NormalBorder()).
		BorderForeground(lipgloss.Color("240")).
		BorderBottom(true).
		Bold(true)
	s.Selected = s.Selected.
		Foreground(lipgloss.Color("229")).
		Background(lipgloss.Color("57")).
		Bold(false)
	t.SetStyles(s)

	m := model{
		table:        t,
		config:       cfg,
		sshManager:   sshMgr,
		gitManager:   gitMgr,
		shellManager: shellMgr,
		emptyStartup: emptyStartup,
	}

	// Load initial data or show welcome
	if emptyStartup {
		m.mode = viewTable
		m.statusMsg = "👋 Welcome! Press 'n' to add an account or 's' to auto-sync from existing setup"
	} else {
		m.statusMsg = fmt.Sprintf("Loaded %d account(s) • Press 'a' to apply configs", len(cfg.ListAccounts()))
	}

	m = m.refreshTable()

	p := tea.NewProgram(m, tea.WithAltScreen())
	if _, err := p.Run(); err != nil {
		return fmt.Errorf("error running program: %w", err)
	}

	return nil
}
